inverse/beamformer_lcmv.m

```diff
@@ -191,143 +205,189 @@
-        eta = u(:,1);
-        lf  = lf * eta;
-        if ~isempty(subspace), lforig = lforig * eta; end
-        dipout.ori{i} = eta;
+          % and compute the leadfield for that orientation
+          lf  = lf * maxpowori;
+          dipout.ori{i} = maxpowori;
+          dipout.eta{i} = eta;
+          if hassubspace, lforig = lforig * maxpowori; end
+      end
     end
-  end
   
-  if hasfilter
-    % use the provided filter
-    filt = dip.filter{i};
-  elseif strcmp(weightnorm,'nai')
-    % Van Veen's Neural Activity Index
-    % below equation is equivalent to following:  
-    % filt = pinv(lf' * invCy * lf) * lf' * invCy; 
-    % filt = filt/sqrt(noise*filt*filt');
-    filt = pinv(sqrt(noise * lf' * invCy_squared * lf)) * lf' *invCy; % based on Sekihara & Nagarajan 2008 eqn. 4.15
-  elseif strcmp(weightnorm,'unitnoisegain')
-    % Unit-noise gain minimum variance (aka Borgiotti-Kaplan) beamformer
-    % below equation is equivalent to following:  
-    % filt = pinv(lf' * invCy * lf) * lf' * invCy; 
-    % filt = filt/sqrt(filt*filt');
-    filt = pinv(sqrtm(lf' * invCy_squared * lf)) * lf' *invCy;     % Sekihara & Nagarajan 2008 eqn. 4.15
-  else
     % construct the spatial filter
-    filt = pinv(lf' * invCy * lf) * lf' * invCy;              % van Veen eqn. 23, use PINV/SVD to cover rank deficient leadfield
+    switch weightnorm
+      case 'nai'
+        % Van Veen's Neural Activity Index
+        % below equation is equivalent to following:  
+        % filt = pinv(lf' * invC * lf) * lf' * invC; 
+        % filt = filt/sqrt(noise*filt*filt'); 
+        % the scaling term in the denominator is sqrt of projected noise, as per eqn. 2.67 of Sekihara & Nagarajan 2008 (S&N)
+        if fixedori
+          filt = pinv(sqrt(noise * lf' * invC_squared * lf)) * lf' *invC; % based on S&N eqn. 4.08
+        else
+          ft_error('vector version of nai weight normalization is not implemented');
+        end
+      case 'unitnoisegain'
+        % filt*filt' = I
+        % Unit-noise gain minimum variance (aka Borgiotti-Kaplan) beamformer
+        % below equation is equivalent to following:  
+        % filt = pinv(lf' * invC * lf) * lf' * invC; 
+        % filt = filt/sqrt(filt*filt');
+        if fixedori
+          filt = pinv(sqrt(lf' * invC_squared * lf)) * lf' *invC; % S&N eqn. 4.15
+        else
+          % compute the matrix that is used for scaling of the filter's rows, as per eqn. 4.83
+          denom = pinv(lf' * invC * lf);
+          gamma = denom * (lf' * invC_squared * lf) * denom;
+          
+          % compute the spatial filter, as per eqn. 4.85
+          filt = diag(1./sqrt(diag(gamma))) * denom * lf' * invC;
+        end
+      case 'arraygain'
+        % filt*lf = ||lf||, applies to scalar leadfield, and to one of the possibilities of the vector version, eqn. 4.75
+        lfn  = lf./norm(lf);
+        filt = pinv(lfn' * invC * lfn) * lfn' * invC; % S&N eqn. 4.09
+ 
```

Comment:
Does this hold in the vector case? I tried a quick check in MATLAB and I don't think the constraints (eq. 4.63-4.65 in Sekihara & Nagarajan 2008) are fulfilled.

I checked the weightnorm vector case as well (both against book and quick example) and that looks good to me!

Reply:
so your final verdict is that this is resolved, right?

Reply:
The weight norm vector case yes.
About the array gain vector case I am not sure (that's where I referred to the equations).

I think we are using different terminology and I wasn't clear in my last comment, sorry! For me weight norm = unit noise gain, but here weight norm is the option that can _also_ carry array gain.

So I checked unit noise gain vector and that looks alright (both compared to formula and checking constraints on toy example) , but when I checked array gain vector as it is implemented above, I am not sure the constraints hold.

Reply:
Ah, I see. Thanks for the clarification. Indeed weightnorm is used here as the option name, that implements the different flavours of weight normalisation. This naming was not invented by me. 
Anyway, the arraygain normalisation for the vector case is taken from equation 4.75

Reply:
Ah okay, I assumed the first option for vector beamformers. For this I can confirm that the _correct_ constraint (now eq. 4.74) holds for my toy example :) 
Maybe you can add the vector case equation number to the comment? (Right now it only has the scalar case eq. number -- might save future confusion over the different vector formulas, might also just be nitpicky.)

Reply:
I like nitpicky, yet the formula number is already there: in line 362

Reply:
Ha, I overlooked that with `% S&N eqn. 4.09` being the comment right behind the implementation - my bad!!